llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp

```diff
@@ -2116,7 +2116,9 @@ struct DSEState {
       return true;
 
     if (auto *LoadI = dyn_cast<LoadInst>(Store->getOperand(0))) {
-      if (LoadI->getPointerOperand() == Store->getOperand(1)) {
+      if (LoadI->getPointerOperand() == Store->getOperand(1) ||
+          AA.isMustAlias(MemoryLocation::get(LoadI),
```

Comment:
Confirming my understanding of the problem: checking that the address spaces are not equal is insufficient because address spaces may not be disjoint, correct?